File: tests/blake2/blake2b.c
// blake2b.c
// A simple BLAKE2b Reference Implementation.

#include "blake2b.h"
#include <stdio.h>
#include <string.h>


// Cyclic right rotation.

#ifndef ROTR64
#define ROTR64(x, y)  (((x) >> (y)) ^ ((x) << (64 - (y))))
#endif

// Little-endian byte access.

#define B2B_GET64(p)                            \
    (((uint64_t) ((uint8_t *) (p))[0]) ^        \
    (((uint64_t) ((uint8_t *) (p))[1]) << 8) ^  \
    (((uint64_t) ((uint8_t *) (p))[2]) << 16) ^ \
    (((uint64_t) ((uint8_t *) (p))[3]) << 24) ^ \
    (((uint64_t) ((uint8_t *) (p))[4]) << 32) ^ \
    (((uint64_t) ((uint8_t *) (p))[5]) << 40) ^ \
    (((uint64_t) ((uint8_t *) (p))[6]) << 48) ^ \
    (((uint64_t) ((uint8_t *) (p))[7]) << 56))

// G Mixing function.

#define B2B_G(a, b, c, d, x, y) {   \
    v[a] = v[a] + v[b] + x;         \
    v[d] = ROTR64(v[d] ^ v[a], 32); \
    v[c] = v[c] + v[d];             \
    v[b] = ROTR64(v[b] ^ v[c], 24); \
    v[a] = v[a] + v[b] + y;         \
    v[d] = ROTR64(v[d] ^ v[a], 16); \
    v[c] = v[c] + v[d];             \
    v[b] = ROTR64(v[b] ^ v[c], 63); }

// Initialization Vector.

static const uint64_t blake2b_iv[8] = {
    0x6A09E667F3BCC908, 0xBB67AE8584CAA73B,
    0x3C6EF372FE94F82B, 0xA54FF53A5F1D36F1,
    0x510E527FADE682D1, 0x9B05688C2B3E6C1F,
    0x1F83D9ABFB41BD6B, 0x5BE0CD19137E2179
};

static const uint8_t sigma[12][16] = {
        { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15 },
        { 14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3 },
        { 11, 8, 12, 0, 5, 2, 15, 13, 10, 14, 3, 6, 7, 1, 9, 4 },
        { 7, 9, 3, 1, 13, 12, 11, 14, 2, 6, 5, 10, 4, 0, 15, 8 },
        { 9, 0, 5, 7, 2, 4, 10, 15, 14, 1, 11, 12, 6, 8, 3, 13 },
        { 2, 12, 6, 10, 0, 11, 8, 3, 4, 13, 7, 5, 15, 14, 1, 9 },
        { 12, 5, 1, 15, 14, 13, 4, 10, 0, 7, 6, 3, 9, 2, 8, 11 },
        { 13, 11, 7, 14, 12, 1, 3, 9, 5, 0, 15, 4, 8, 6, 2, 10 },
        { 6, 15, 14, 9, 11, 3, 0, 8, 12, 2, 13, 7, 1, 4, 10, 5 },
        { 10, 2, 8, 4, 7, 6, 1, 5, 15, 11, 9, 14, 3, 12, 13, 0 },
        { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15 },
        { 14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3 }
    };


void printv(char* n, int len, uint64_t v[]) {
  printf("%s [",n);
    for (int i = 0; i < len; i++) {
      if (i>0) printf(",");
      printf("%llu",v[i]);
    }
  printf("]");
}

void printb(char* n, int len, uint8_t v[]) {
  printf("%s [",n);
    for (int i = 0; i < len; i++) {
      if (i>0) printf(",");
      printf("%d",v[i]);
    }
  printf("]");
}


void printctx(blake2b_ctx ctx) {
  printf("ctx: Blake2BCtx {");
  printb("_b=pack",128,ctx.b);
  printv(",_h=mk",8,ctx.h);
  printf(",_t0=%llu,_t1=%llu,_c=%lu,_outlen=%lu}\n",ctx.t[0],ctx.t[1],ctx.c,ctx.outlen);
}

// Compression function. "last" flag indicates last block.

static void blake2b_compress(blake2b_ctx *ctx, int last)
{

    int i;
    uint64_t v[16], m[16];

    for (i = 0; i < 8; i++) {           // init work variables
        v[i] = ctx->h[i];
        v[i + 8] = blake2b_iv[i];
    }

    v[12] ^= ctx->t[0];                 // low 64 bits of offset
    v[13] ^= ctx->t[1];                 // high 64 bits
    if (last)                           // last block flag set ?
        v[14] = ~v[14];
    //printv("v1",v);
    for (i = 0; i < 16; i++)            // get little-endian words
        m[i] = B2B_GET64(&ctx->b[8 * i]);
    //printv("m",m);
    for (i = 0; i < 12; i++) {          // twelve rounds
        B2B_G( 0, 4,  8, 12, m[sigma[i][ 0]], m[sigma[i][ 1]]);
        B2B_G( 1, 5,  9, 13, m[sigma[i][ 2]], m[sigma[i][ 3]]);
        B2B_G( 2, 6, 10, 14, m[sigma[i][ 4]], m[sigma[i][ 5]]);
        B2B_G( 3, 7, 11, 15, m[sigma[i][ 6]], m[sigma[i][ 7]]);
        B2B_G( 0, 5, 10, 15, m[sigma[i][ 8]], m[sigma[i][ 9]]);
        B2B_G( 1, 6, 11, 12, m[sigma[i][10]], m[sigma[i][11]]);
        B2B_G( 2, 7,  8, 13, m[sigma[i][12]], m[sigma[i][13]]);
        B2B_G( 3, 4,  9, 14, m[sigma[i][14]], m[sigma[i][15]]);
    }
    //printv("v2",v);
    for( i = 0; i < 8; ++i )
        ctx->h[i] ^= v[i] ^ v[i + 8];
}

// Initialize the hashing context "ctx" with optional key "key".
//      1 <= outlen <= 64 gives the digest size in bytes.
//      Secret key (also <= 64 bytes) is optional (keylen = 0).

int blake2b_init(blake2b_ctx *ctx, size_t outlen,
    const void *key, size_t keylen)        // (keylen=0: no key)
{
    size_t i;

    if (outlen == 0 || outlen > 64 || keylen > 64)
        return -1;                      // illegal parameters

    for (i = 0; i < 8; i++)             // state, "param block"
        ctx->h[i] = blake2b_iv[i];
    //printf("ctxh %llu -> ",ctx->h[0]);
    ctx->h[0] ^= 0x01010000 ^ (keylen << 8) ^ outlen;
    //printf("%llu\n",ctx->h[0]);

    ctx->t[0] = 0;                      // input count low word
    ctx->t[1] = 0;                      // input count high word
    ctx->c = 0;                         // pointer within buffer
    ctx->outlen = outlen;

    for (i = keylen; i < 128; i++)      // zero input block
        ctx->b[i] = 0;
    if (keylen > 0) {
        blake2b_update(ctx, key, keylen);
        ctx->c = 128;                   // at the end
    }


    return 0;
}

// Add "inlen" bytes from "in" into the hash.

void blake2b_update(blake2b_ctx *ctx,
    const void *in, size_t inlen)       // data bytes
{
    size_t i;

    for (i = 0; i < inlen; i++) {
        if (ctx->c == 128) {            // buffer full ?
            ctx->t[0] += ctx->c;        // add counters
            if (ctx->t[0] < ctx->c)     // carry overflow ?
                ctx->t[1]++;            // high word
            blake2b_compress(ctx, 0);   // compress (not last)
            ctx->c = 0;                 // counter to zero
        }
        ctx->b[ctx->c++] = ((const uint8_t *) in)[i];
    }
}

// Generate the message digest (size given in init).
//      Result placed in "out".

void blake2b_final(blake2b_ctx *ctx, void *out)
{
    size_t i;

    ctx->t[0] += ctx->c;                // mark last block offset
    if (ctx->t[0] < ctx->c)             // carry overflow
        ctx->t[1]++;                    // high word

    while (ctx->c < 128)                // fill up with zeros
        ctx->b[ctx->c++] = 0;
    blake2b_compress(ctx, 1);           // final block flag = 1

    // little endian convert and store
    for (i = 0; i < ctx->outlen; i++) {
        ((uint8_t *) out)[i] =
            (ctx->h[i >> 3] >> (8 * (i & 7))) & 0xFF;
    }
}

// Convenience function for all-in-one computation.

int blake2b(void *out, size_t outlen,
    const void *key, size_t keylen,
    const void *in, size_t inlen)
{
    blake2b_ctx ctx;

    if (blake2b_init(&ctx, outlen, key, keylen))
        return -1;
    printf("init: outlen=%lu keylen=%lu\n",outlen,keylen);
    printf("pre-update_");
    printctx(ctx);
    blake2b_update(&ctx, in, inlen);
    printf("post-update_");
    printctx(ctx);
    blake2b_final(&ctx, out);

    return 0;
}





static void selftest_seq(uint8_t *out, size_t len, uint32_t seed)
{
    size_t i;
    uint32_t t, a , b;

    a = 0xDEAD4BAD * seed;              // prime
    b = 1;

    for (i = 0; i < len; i++) {         // fill the buf
        t = a + b;
        a = b;
        b = t;
        out[i] = (t >> 24) & 0xFF;
    }
}

int blake2b_selftest() {
    // grand hash of hash results
    const uint8_t blake2b_res[32] = {
        0xC2, 0x3A, 0x78, 0x00, 0xD9, 0x81, 0x23, 0xBD,
        0x10, 0xF5, 0x06, 0xC6, 0x1E, 0x29, 0xDA, 0x56,
        0x03, 0xD7, 0x63, 0xB8, 0xBB, 0xAD, 0x2E, 0x73,
        0x7F, 0x5E, 0x76, 0x5A, 0x7B, 0xCC, 0xD4, 0x75
    };
    // parameter sets
    const size_t b2b_md_len[4] = { 20, 32, 48, 64 };
    const size_t b2b_in_len[6] = { 0, 3, 128, 129, 255, 1024 };

    size_t i, j, outlen, inlen;
    uint8_t in[1024], md[64], key[64];
    blake2b_ctx ctx;

    // 256-bit hash for testing
    if (blake2b_init(&ctx, 32, NULL, 0))
        return -1;

    for (i = 0; i < 4; i++) {
        outlen = b2b_md_len[i];
        for (j = 0; j < 6; j++) {
            inlen = b2b_in_len[j];
            printf("outlen: %lu,inlen: %lu\n",outlen,inlen);
            selftest_seq(in, inlen, inlen);     // unkeyed hash
            printb("in:",inlen,in);printf("\n");
            blake2b(md, outlen, NULL, 0, in, inlen);
            printb("bl1:",outlen,md);printf("\n");
            blake2b_update(&ctx, md, outlen);   // hash the hash
            printctx(ctx);
            selftest_seq(key, outlen, outlen);  // keyed hash
            printb("key:",outlen,key);printf("\n");
            blake2b(md, outlen, key, outlen, in, inlen);
            printb("bl2:",outlen,md);printf("\n");
            blake2b_update(&ctx, md, outlen);   // hash the hash
            printctx(ctx);
        }
    }

    // compute and compare the hash of hashes
    blake2b_final(&ctx, md);
    for (i = 0; i < 32; i++) {
        if (md[i] != blake2b_res[i])
            return -1;
    }

    return 0;
}


int main() {

  // printf("rotr64 31 47 = %llu\n",ROTR64((uint64_t)31,47));
  uint64_t v[16] = {7640891576939301192llu,13503953896175478587llu,4354685564936845355llu,11912009170470909681llu,5840696475078001361llu,11170449401992604703llu,2270897969802886507llu,6620516959819538809llu,7640891576956012808llu,13503953896175478587llu,4354685564936845355llu,11912009170470909681llu,5840696475078001361llu,11170449401992604703llu,2270897969802886507llu,6620516959819538809llu};
    uint64_t m[16] = {6513249llu,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};

  /*for (int i = 0; i < 16; i++) {
      v[i] = i;
      m[i] = i;
      }*/

    // printf("sigs: %llu %llu\n",m[sigma[0][ 0]], m[sigma[0][ 1]]);
        B2B_G( 0, 4,  8, 12, m[sigma[0][ 0]], m[sigma[0][ 1]]);
        B2B_G( 1, 5,  9, 13, m[sigma[0][ 2]], m[sigma[0][ 3]]);
        B2B_G( 2, 6, 10, 14, m[sigma[0][ 4]], m[sigma[0][ 5]]);
        B2B_G( 3, 7, 11, 15, m[sigma[0][ 6]], m[sigma[0][ 7]]);
        B2B_G( 0, 5, 10, 15, m[sigma[0][ 8]], m[sigma[0][ 9]]);
        B2B_G( 1, 6, 11, 12, m[sigma[0][10]], m[sigma[0][11]]);
        B2B_G( 2, 7,  8, 13, m[sigma[0][12]], m[sigma[0][13]]);
        B2B_G( 3, 4,  9, 14, m[sigma[0][14]], m[sigma[0][15]]);

       // printv("v",v);
    blake2b_ctx ctx;

    // 256-bit hash for testing
    if (blake2b_init(&ctx, 32, NULL, 0))
      return -1;
    /*
    ctx.b[0]='a';
    ctx.b[1]='b';
    ctx.b[2]='c';
    */
    printctx(ctx);

    // blake2b_compress(&ctx,0);
    char* d = "akjsvhsjkdfhvjsakhfvjkdhsvnfklsjanvskdjvkdjsvfksdhnvkjdfavksnksjnfajvjklsnkjsfdajkldahvfadvndfjvalkhfnvkdfjsnvkljnkfkajsdnaskjskdjvfsfvkkdsadshfjkalsvaklsdhvnlksfjvnskdhnhjdnshsdfkvnfvads";
    printf ("length: %lu\n", strlen(d));
    blake2b_update(&ctx,d,strlen(d));
    uint8_t o[32];
    blake2b_final(&ctx,o);
    printctx(ctx);
    printf("o: [");
    for (int i = 0 ; i < 32; i ++) {
      printf("%d,",o[i]);
    }
    printf("]\n");


    printf("test: %d", blake2b_selftest());

  return 0;
}

File: src/Pact/Types/Blake2b.hs
{-# LANGUAGE OverloadedStrings #-}

-- |
-- Module      :  Pact.Types.PureBlake2
-- Copyright   :  (C) 2016 Stuart Popejoy
-- License     :  BSD-style (see the file LICENSE)
-- Maintainer  :  Stuart Popejoy <stuart@kadena.io>
--
-- Pure implementation of BLAKE2b for use in GHCJS.
-- Adapted from C code in https://github.com/mjosaarinen/blake2_mjosref
--

module Pact.Types.Blake2b where

import Data.Word
import Data.Bits
import Data.ByteString as B (ByteString,index,unpack,pack,length,take,splitAt,null,replicate)
import Data.Vector as V (Vector,(//),(!),fromList,(++))
import Control.Lens ((&),(<&>))
import Prelude as P hiding (last)
import Data.Monoid


type V = Vector

-- | state context
data Blake2BCtx = Blake2BCtx
  { _b :: ByteString -- 128 input buffer
  , _h :: V Word64        --   8 chained state
  , _t0 :: Word64         --   total number of bytes 0
  , _t1 :: Word64         --                         1
  , _c :: Int        -- pointer for b[]
  , _outlen :: Int   -- digest size
  } deriving (Eq)
instance Show Blake2BCtx where
  show (Blake2BCtx b h t0 t1 c outlen) =
    "Blake2BCtx {_b=pack" P.++ show (unpack b) P.++ ", _h=mk" P.++ show h P.++
    ", _t0=" P.++ show t0 P.++ ", _t1=" P.++ show t1 P.++ ", _c=" P.++ show c P.++
    ", _outlen=" P.++ show outlen P.++ "}"


at :: V a -> Int -> a
at v i = v ! i


upd :: Int -> (V a -> a -> a) -> V a -> V a
upd i f v = v // [(i,f v (v ! i))]


mk :: [a] -> V a
mk = fromList

cat :: V Word64 -> V Word64 -> V Word64
cat = (V.++)

-- | flipped foldl
ffoldl :: Foldable t => b -> t a -> (b -> a -> b) -> b
ffoldl i l f = foldl f i l


-- | Cyclic right rotation.
rotr64 :: Word64 -> Int -> Word64
rotr64 x y = (x `shiftR` y) `xor` (x `shiftL` (64 - y))

-- | Little-endian byte access.

b2b_get64 :: Int -> ByteString -> Word64
b2b_get64 o p =
  let ix64 i = fromIntegral $ index p (o+i)
  in
        ix64 0               `xor`
       (ix64 1  `shiftL` 8 ) `xor`
       (ix64 2  `shiftL` 16) `xor`
       (ix64 3  `shiftL` 24) `xor`
       (ix64 4  `shiftL` 32) `xor`
       (ix64 5  `shiftL` 40) `xor`
       (ix64 6  `shiftL` 48) `xor`
       (ix64 7  `shiftL` 56)


-- | G Mixing function.
b2b_g :: Int -> Int -> Int -> Int -> Word64 -> Word64 -> V Word64 -> V Word64
b2b_g a b c d x' y' v' =
  v' &
  upd a (\v x -> x + (v `at` b) + x') &
  upd d (\v x -> rotr64 (x `xor` (v `at` a)) 32) &
  upd c (\v x -> x + (v `at` d)) &
  upd b (\v x -> rotr64 (x `xor` (v `at` c)) 24) &
  upd a (\v x -> x + (v `at` b) + y') &
  upd d (\v x -> rotr64 (x `xor` (v `at` a)) 16) &
  upd c (\v x -> x + (v `at` d)) &
  upd b (\v x -> rotr64 (x `xor` (v `at` c)) 63)

-- | Initialization Vector.
blake2b_iv :: V Word64
blake2b_iv = mk [
    0x6A09E667F3BCC908, 0xBB67AE8584CAA73B,
    0x3C6EF372FE94F82B, 0xA54FF53A5F1D36F1,
    0x510E527FADE682D1, 0x9B05688C2B3E6C1F,
    0x1F83D9ABFB41BD6B, 0x5BE0CD19137E2179
    ]

sigma :: V (V Int)
sigma = mk [
        mk [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15 ],
        mk [ 14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3 ],
        mk [ 11, 8, 12, 0, 5, 2, 15, 13, 10, 14, 3, 6, 7, 1, 9, 4 ],
        mk [ 7, 9, 3, 1, 13, 12, 11, 14, 2, 6, 5, 10, 4, 0, 15, 8 ],
        mk [ 9, 0, 5, 7, 2, 4, 10, 15, 14, 1, 11, 12, 6, 8, 3, 13 ],
        mk [ 2, 12, 6, 10, 0, 11, 8, 3, 4, 13, 7, 5, 15, 14, 1, 9 ],
        mk [ 12, 5, 1, 15, 14, 13, 4, 10, 0, 7, 6, 3, 9, 2, 8, 11 ],
        mk [ 13, 11, 7, 14, 12, 1, 3, 9, 5, 0, 15, 4, 8, 6, 2, 10 ],
        mk [ 6, 15, 14, 9, 11, 3, 0, 8, 12, 2, 13, 7, 1, 4, 10, 5 ],
        mk [ 10, 2, 8, 4, 7, 6, 1, 5, 15, 11, 9, 14, 3, 12, 13, 0 ],
        mk [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15 ],
        mk [ 14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3 ]
        ]


-- | Compression function. "last" flag indicates last block.
blake2b_compress :: Blake2BCtx -> Bool -> Blake2BCtx
blake2b_compress ctx last =
  let
    v1 = (_h ctx `cat` blake2b_iv) &
      upd 12 (\_ x -> (x `xor` _t0 ctx)) &                    -- low 64 bits of offset
      upd 13 (\_ x -> (x `xor` _t1 ctx)) &                    -- high 64 bits
      (if last then upd 14 (\_ x -> complement x) else id)     -- last block flag set ?
    m = mk $ (`map` [0..15]) $ \i -> b2b_get64 (8 * i) (_b ctx)         -- get little-endian words
    v2 = ffoldl v1 sigma $ runRound m                        -- 12 rounds
    h = mk $ (`map` [0..7]) $ \i ->
          (_h ctx `at` i) `xor` (v2 `at` i) `xor` (v2 `at` (i + 8))
  in ctx { _h = h }

runRound :: V Word64 -> V Word64 -> V Int -> V Word64
runRound m v s = v &
  b2b_g 0 4  8 12 (m `at` (s `at`  0)) (m `at` (s `at`  1)) &
  b2b_g 1 5  9 13 (m `at` (s `at`  2)) (m `at` (s `at`  3)) &
  b2b_g 2 6 10 14 (m `at` (s `at`  4)) (m `at` (s `at`  5)) &
  b2b_g 3 7 11 15 (m `at` (s `at`  6)) (m `at` (s `at`  7)) &
  b2b_g 0 5 10 15 (m `at` (s `at`  8)) (m `at` (s `at`  9)) &
  b2b_g 1 6 11 12 (m `at` (s `at` 10)) (m `at` (s `at` 11)) &
  b2b_g 2 7  8 13 (m `at` (s `at` 12)) (m `at` (s `at` 13)) &
  b2b_g 3 4  9 14 (m `at` (s `at` 14)) (m `at` (s `at` 15))

-- | Initialize the hashing context "ctx" with optional key "key".
--      1 <= outlen <= 64 gives the digest size in bytes.
--      Secret key (also <= 64 bytes) is optional (keylen = 0).
blake2b_init :: Word64 -> ByteString -> Either String Blake2BCtx
blake2b_init outlen key
  | outlen < 1 || outlen > 64 = Left $ "outlen must be > 0 and <= 64 bytes" P.++ show outlen
  | B.length key > 64 = Left "Key must be <= 64 bytes"
  | otherwise = Right $
    let keylen :: Word64
        keylen = fromIntegral $ B.length key
        iv = blake2b_iv
    in Blake2BCtx
          {
            -- state, "param block"
            _h = upd 0 (\_ x -> x `xor` 0x01010000 `xor` (keylen `shiftL` 8) `xor` outlen) iv
          , _t0 = 0 -- input count low word
          , _t1 = 0 -- input count high word
          , _c = 0  -- pointer within buffer
          , _outlen = fromIntegral outlen
          , _b = B.replicate 128 0 -- zero input block.
          } &
          (\ctx -> if keylen > 0 then blake2b_update key ctx & (\c -> c { _c = 128}) else ctx)

-- | Slice a bytestring into segments: 'slice c l bs' means the first
-- segment will be length c but no longer than l, and the following will
-- be length l, with the last being whatever remains.
slice :: Int -> Int -> ByteString -> [ByteString]
slice c l bs = go (B.splitAt (if c >= l then l else l - c) bs) where
  go (x,y) | B.null y = [x]
           | otherwise = x:go (B.splitAt l y)

-- | Add "inlen" bytes from "in" into the hash.
blake2b_update :: ByteString -> Blake2BCtx -> Blake2BCtx
blake2b_update bs ctx
  | B.null bs = ctx
  | otherwise = ffoldl ctx (slice (_c ctx) 128 bs) $ \cx s ->
      resetMaybe cx & cpyBuf s
  where resetMaybe cx
          | _c cx < 128 = cx
          | otherwise =
              let t0 = _t0 cx + 128
                  t1 = _t1 cx + (if t0 < 128 then 1 else 0)
              in (blake2b_compress (cx { _t0 = t0, _t1 = t1 }) False) { _c = 0 }
        cpyBuf s cx = cx {
          _b = B.take (_c cx) (_b cx) <> s <> B.replicate (128 - _c cx - B.length s) 0,
          _c = B.length s + _c cx
          }

-- | Generate the message digest (size given in init).
blake2b_final :: Blake2BCtx -> ByteString
blake2b_final ctx = finalCompress ctx & conv where
  finalCompress cx =
    let c = fromIntegral $ _c cx
        t0 = _t0 cx + c
        t1 = _t1 cx + (if t0 < c then 1 else 0)
        b = B.take (_c cx) (_b cx) <> B.replicate (128 - _c cx) 0
    in blake2b_compress (cx { _t0 = t0, _t1 = t1, _b = b }) True
  conv cx = B.pack $ (`map` [0 .. pred (_outlen cx)]) $ \i ->
    fromIntegral $ ((_h cx `at` (i `shiftR` 3)) `shiftR` (8 * (i .&. 7))) .&. 0xFF

-- | All together now! 'blake2b outlen key inB' hashes inB, with
-- optional key (empty means no key), to length outlen.
blake2b :: Int -> ByteString -> ByteString -> Either String ByteString
blake2b outlen key inB =
  blake2b_init (fromIntegral outlen) key <&>
  (blake2b_final . blake2b_update inB)

-- Test value
_ctx32 :: Blake2BCtx
_ctx32 = let Right c = blake2b_init 32 mempty in c
